### DISCUSSION TITLE
Midi DIN in issue

### DISCUSSION THREAD

**Discussion by gitrimius** [shanteacontrols/OpenDeck#83] [Q&A]
Hi,
I am encountering something that I can not explain. I did a few tests to diagnose.
My setup is as follows:

- STM32F411 board with custom fw (based on 5.5.2), receiving midi on pins with the circuit noted in the wiki.
- Beatstep Pro sequencer (or other midi source) => midi DIN in on STM.
- Leds receiving notes (single value) from DIN.

result:
- at slow speeds everything seems to work fine. The leds make a sequence 1,2,3,4,5,6,7,8 and so on.
- at higher speeds (16th notes at +55bpm) things start to fail. Leds are skipped or stay lid. So I guess midi note on/off messages are missed etc…
All this is the result from Beatstep Pro to Midi Din in.
At the OLED screen notes appear of >127 (for example 149), also velocities can range >127.
see following movie: https://youtu.be/kfhkq4ZfvjE

So I thought something is wrong with my Midi in/out circuit. I proceeded with the following setup:
- Beatstep Pro => via USB midi to STM32
Everything works fine even at very high bpm.

Next test was:
- Forward all Midi DIN in data to USB (by setting Midi Merge in the web-interface) and monitoring on the PC.
Data looks fine. So no Velocities of >127, nor notes of >127, but still leds are skipped or stay lid and at the OLED screen notes appear of >127 (for example 149), also velocities can range >127.
See following movie: https://youtu.be/RXY2JhTaEAg

Since the monitoring on PC from the forwarded data (from DIN to USB) seems to be ok (but the leds are not reacting correspondingly and the OLED shows strange values), I wonder whether there may be something wrong in the firmware?
Would you have any suggestions how to proceed my testing?

Thanks a lot!

**Comment by paradajz** [comment]
I'll do some tests later today.

**Reply by paradajz** [reply]
In the meantime try disabling the display and check if there are any changes. I haven't limited the amount of writes to the screen which could be a problem.

**Reply by gitrimius** [reply]
okay, I'll try this out this evening.

**Reply by paradajz** [reply]
Actually I *have* limited the writes, but only to 10ms. Anyways try disabling the screen and let me know.

**Reply by gitrimius** [reply]
I just tried. Disabling the OLED actually has effect on
- the rate by which the leds change.
- the error rate is much lower.
But it still misses note on/off quite a bit and I can not speed up to higher speeds.

**Comment by paradajz** [comment]
This cannot be solved easily or any time soon. The only way to solve this is to port the entire OpenDeck firmware to some RTOS, which is a very large undertaking. Sorry.

**Comment by paradajz** [comment]
I've just pushed some performance optimizations to master so you can try that out. Haven't yet figured the issues with display. Make sure to also pull new versions of submodules.

**Reply by gitrimius** [reply]
Hi, just got back from an offline weekend.
I'll give it a test this evening,
tnx!

**Reply by gitrimius** [reply]
Hi, I just tested it and it seems to work great. At the speeds at which I need to drive the leds all note messages seems to be transferred correctly. When I enable the OLED there seems to be a small visible difference in the timing when I try to enable 2 leds at the same time, but no messages are skipped etc. 
Thanks a lot for that!

**Reply by paradajz** [reply]
Great. I'll see what I can do regarding display optimization in the future.
